src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php

```diff
@@ -176,43 +236,49 @@ public function indexAction(Request $request)
      */
     public function createAction(Request $request)
     {
-        $this->isGrantedOr403('create');
+        $configuration = $this->configurationFactory->create($this->metadata, $request);
 
-        $resource = $this->createNew();
-        $form = $this->getForm($resource);
+        $this->isGrantedOr403($configuration, ResourceActions::CREATE);
 
-        if ($request->isMethod('POST') && $form->submit($request)->isValid()) {
-            $resource = $this->domainManager->create($form->getData());
+        $resource = $this->createNew($configuration);
+        $form = $this->createForm($configuration, $resource);
 
-            if ($this->config->isApiRequest()) {
-                if ($resource instanceof ResourceEvent) {
-                    throw new HttpException($resource->getErrorCode(), $resource->getMessage());
-                }
+        if ($request->isMethod('POST') && $form->submit($request)->isValid()) {
+            $this->eventDispatcher->dispatch(GenericResourceEvents::PRE_CREATE, new GenericResourceEvent($resource, $this->metadata, $configuration, ResourceActions::CREATE));
+            $event = $this->eventDispatcher->dispatchResourceEvent(ResourceEvents::PRE_CREATE, $resource);
 
-                return $this->handleView($this->view($resource, 201));
+            if ($event->isPropagationStopped() && !$configuration->isHtmlRequest()) {
+                throw new HttpException($event->getErrorCode(), $event->getMessage());
             }
 
-            if ($resource instanceof ResourceEvent) {
-                return $this->redirectHandler->redirectToIndex();
+            $this->manager->persist($resource);
```

Comment:
Yes, but the thing is that I do not want to encourage firing this events outside of controller. If you want to really be sure that something happens every time entity is created/updated/deleted - you need to use Doctrine events. That's the sad truth. Otherwise you will need to remember about these events and inject dispatcher into weird spots.

Basically, these events are controller specific and you should fire them only in controller. For other use cases - use doctrine events specific to your driver. That's why we have them separated from the manager. Manager is used everywhere.